draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,206 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+
+* `PingPongInitialize(prep_share)` - Sent by the Leader to the Helper in the
+  initial request. The payload includes its first-round prep share.
+
+* `PingPongContinue(prep_msg, prep_share)` - Sent by the Leader or Helper in an
+  intermediate request or response. The payload includes the prep message from
+  the current round and the sender's prep share for the next round.
+
+* `PingPongFinish(prep_msg)` - Sent by the Leader in the last request
+  (even-round VDAF) or the Helper in the last response (odd-round VDAF). The
+  payload is the final-round prep message.
+
+* `PingPongFinished()` - Sent by the Helper in the final response in even-round
+  VDAFs. There is no payload.
+
+Each Aggregator handles these messages according to the following transition rule,
+so long as they are in the `Continued` state:
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   state: State,
+                   inbound: Optional[Message],
+                   ) -> (State, Optional[Message]):
+    if type(inbound) == PingPongInitialize:
+        prep_msg, peer_prep_share = None, inbound.prep_share
+    elif type(inbound) == PingPongContinue:
+        prep_msg = inbound.prep_msg
+        peer_prep_share = inbound.prep_share
+    elif type(inbound) == PingPongFinish:
+        prep_msg, peer_prep_share = inbound.prep_msg, None
+    elif type(inbound) == PingPongFinished:
+        return (Rejected(), None)
+    else: # inbound == None
+        prep_msg, peer_prep_share = None, None
+
+    try:
+        # If `prep_msg == None` then this is the start of the
+        # first round. Otherwise, `prep_msg` is the prep message
+        # from the end of the previous round.
+        out = Vdaf.prep_next(host_prep_state, prep_msg)
+        if type(out) == Vdaf.OutShare:
+            # The Leader reaches this state upon processing a
+            # response from the Helper, so there is no outbound
+            # message.
+            outbound = None if is_leader else PingPongFinished()
+            return (Finished(out), outbound)
+        (host_prep_state, host_prep_share) = out
+
+        if peer_prep_share != None:
+            prep_shares = [peer_prep_share, host_prep_share]
+            if is_leader:
+                prep_shares.reverse()
+            prep_msg = Vdaf.prep_shares_to_prep(
+                agg_param,
+                prep_shares,
+            )
+            out = Vdaf.prep_next(host_prep_state, prep_msg)
+            if type(out) == Vdaf.OutShare:
+                return (Finished(out), PingPongFinish(prep_msg))
+            (host_prep_state, host_prep_share) = out
+
+        if is_leader and inbound == None:
+            outbound = PingPongInitialize(host_prep_share)
+        else:
+            outbound = PingPongContinue(prep_msg, host_prep_share)
+        return (Continued(host_prep_state), outbound)
+    except:
+        return (Rejected(), None)
```

Comment:
WDYT about introducing a `PingPongRejected` message instead of returning `None`?

Other than being more explicit about the meaning of this return value, the way things are currently written I suppose someone could misinterpret things such that the other aggregator would continue to call `ping_pong_next` with `inbound == None`, which is not what we want.

Reply:
When would that happen? If I'm in the Rejected state, I'm not going to keep processing messages for that report.

Reply:
This aggregator would be in `Rejected` state, but the other aggregator wouldn't be in `Rejected` yet. By returning `None` here, someone misinterpreting this text might think that this means they need to send "no message" to the other aggregator, which would then call `ping_pong_next` with `inbound == None`, which would cause the peer aggregator to misbehave.

But this is just my guess at how things might go wrong. My bigger concern is being explicit: `PingPongRejected` has a more explicit meaning than `None`, and will make it easier for protocols using these functions to specify handling.

Reply:
Hmm, let me think about this. We get to `Rejected()` in three ways:
1. prep_init() raises an error
2. prep_next() or prep_shares_to_prep() raises an error
3. ping_pong_next() is called on `PingPongFinished()`

Whether there is an explicit message sent in DAP kind of depends: in cases (1) and (2) only the Helper sends an explicit message, at least for now (this might change with https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/438). Case (3) should actually be unreachable if the ping pong API is used correctly.

I think if we want to be more explicit, we should be consistent. I.e., in case (1) there should be an explicit outbound message. But the fact that whether there is an explicit message depends on protocol details, I think I'd prefer to leave the type as `None` and leave it to the protocol to sort out.

Reply:
Good point; one way to resolve this, but perhaps too slowly for this PR, would be to take https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/244 -- then ping-pong initialization could be similarly fused such that `ping_pong_start` would return a state & outbound message, which could then be `PingPongRejected`.

Another option would be not to address this at all: I updated https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/393 to use the ping-pong API from this PR, rejection handling isn't too awkward (and not all of the awkwardness would be resolved even if we took this suggestion).

Reply:
I'd prefer the second option if it's not too awkward. I just saw your changes to 393 and am reviewing now.